### ISSUE TITLE
Bug: ast error parsing generic parameter

### ISSUE THREAD

**Issue by rxliuli** [typescript-eslint/typescript-eslint#5424]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3Abug) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/docs/linting/troubleshooting) and my problem is not listed.

### Relevant Package

typescript-estree

### Playground Link

https://typescript-eslint.io/play/#ts=4.7.2&sourceType=module&showAST=es&code=GYVwdgxgLglg9mABMAPAFQHwAoCGAuRNASgLUQG8BYAKEUQCcBTKEepHGgXyA&eslintrc=N4KABGBEBOCuA2BTAzpAXGYBfEWg&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA

### Repro Code

```typescript
function f<T>(a: T): T {
  return a
}
```

### ESLint Config

```javascript
module.exports = {
  parser: "@typescript-eslint/parser",
  rules: {
    "@typescript-eslint/<rule-name>": ["error", ...<options>],
  },
};
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    // ...
  }
}
```

### Expected Result

ast for generic parameters

```json
{
  "type": "TSTypeParameter",
  "name": "T"
}
```

### Actual Result

the ast of the actual generic parameter

```json
{
  "type": "TSTypeParameter",
  "name": {
    "type": "Identifier",
    "name": "T"
  }
}
```

### Additional Info

The reason I noticed this problem was because while doing a code generator, using eslint as the parser for ts, I found that the generic parameter was not expected, I investigated @babel/parser's and confirmed that the problem is typescript's ast alone some. Maybe this is normal, but it has some friction with the ast-types api, e.g. the result of parsing looks like it should use `b.tsTypeParameter(b.identifier('T'))` to generate an ast for generic parameters, but `b.tsTypeParameter('T')` is actually required, but that's not what the parsing code gets.

[@babel/parser-7.18.0](https://babel.dev/docs/en/next/babel-parser)
![image](https://user-images.githubusercontent.com/24560368/183011315-b3978c26-5bbb-4f31-b62e-3c3e415550bc.png)

[@typescript-eslint/parser-5.30.3](https://typescript-eslint.io/)
![image](https://user-images.githubusercontent.com/24560368/183011361-da2912e5-31e0-4a19-9eb8-35f7a2690394.png)

### Versions

| package                                | version |
| -------------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin`     | `X.Y.Z` |
| `@typescript-eslint/parser`            | `X.Y.Z` |
| `@typescript-eslint/scope-manager`     | `X.Y.Z` |
| `@typescript-eslint/typescript-estree` | `5.32.0` |
| `@typescript-eslint/type-utils`        | `X.Y.Z` |
| `@typescript-eslint/utils`             | `X.Y.Z` |
| `TypeScript`                           | `4.5.2` |
| `ESLint`                               | `X.Y.Z` |
| `node`                                 | `16.14.2` |

**Comment by bradzacher** [comment]
Note that you need to ensure you're using babel configured for TS parsing. The AST you're seeing is the AST for flow code, not TS! I can tell this because the TS-estree ast spec we define prefixes all TS-specific nodes with `TS*` in order to disambiguate the nodes from other languages.

---

There is no problem here from our end - this is the expected AST as it has been defined since its inception!

We are the canonical implementation for the TS-estree ast - plugins build tools and test against our parser, as we are the only parser that can provide type information in addition to the AST.

Whilst babel does parse TS code into an estree-compatible AST, unfortunately their AST is not truly compatible with our TS-estree ast spec.
In the early days babel did their own thing when defining nodes - so there is a lot of nodes that are not properly aligned, sadly. 

This look like one case where babel is not aligned with us, unfortunately.

**Comment by rxliuli** [comment]
That might be an error in the ast-types api, because its definition here requires a string instead of `Identifier`, and the name starts with `TS*`

```ts
export interface TSTypeParameterBuilder {
    (name: string, constraint?: K.TSTypeKind | undefined, defaultParam?: K.TSTypeKind | undefined): namedTypes.TSTypeParameter;
    from(params: {
        comments?: K.CommentKind[] | null;
        constraint?: K.TSTypeKind | undefined;
        default?: K.TSTypeKind | undefined;
        loc?: K.SourceLocationKind | null;
        name: string;
        optional?: boolean;
        typeAnnotation?: K.TypeAnnotationKind | K.TSTypeAnnotationKind | null;
    }): namedTypes.TSTypeParameter;
}
```

**Comment by bradzacher** [comment]
That's not our API. That's a babel api.

The API is correct as per their AST representation, which does not match our representation.
